google/cloud/storage/well_known_parameters.h

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018 Google LLC
```

Comment:
Thinking forward a bit, how do you intend on extending this for representing the possible parameters of other requests?  Some requests have unique preconditions and query parameters, so in order to reuse these **and** allow specifying the parameters unique to each specific request, we'd have to do more mixing of WellKnownParameters and SomeOtherClass.  A great example of something with a **slew** of unique preconditions/filters is an object rewrite request:
https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite

Maybe I'm just not up to speed on C++ style/idioms these days, but it seems like this will quickly become a syntax soup just to support variadic templates.  Do you think it's worth the additional code complexity to do so, as opposed to a builder pattern? E.g.:

    request.applyFilters(FilterType1(1), FilterType2(2));
    // vs
    request.setFilterOfType1(FilterType1(1)).setFilterOfType2(FilterType2(2));

From a maintenance standpoint, the latter:
*  seems a bit easier to grok when we will inevitably need to debug, especially to those without an encyclopedic knowledge of C++ syntax
*  is pretty straightforward since each of these corresponds to a header or query parameter that can only have one value

Honestly, I'm not sure how much simpler the builder pattern would be, given the considerations above.  Do you think the pros of variadic template support outweigh the cons?

Reply:
> Thinking forward a bit, how do you intend on extending this for representing the possible
> parameters of other requests? 

By adding all the well-known parameters to the `WellKnownParameters` class.  Note that this class is not intended for direct use.  An example of the API application developers would use is:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/pull/685/files#diff-2e4f3c4914ea11283c4ce2b05ea34320R65

Or:

```C++
auto metadata = bucket->GetMetadata(IfMetadataMatch(...), UserProject(...)
   /* more parameters if they want */);
```

Or (if we decided to go with just a `Client` class and not `Bucket` and `Object` wrappers):

```C++
GetMetadataRequest request;
request.set_bucket_name("foo");
request.Apply(IfMetadataMatch(...), UserProject(...)
   /* more parameters if they want */);
auto metadata = client->GetBucketMetadata(request);
```

> Maybe I'm just not up to speed on C++ style/idioms these days, but it seems like this
> will quickly become a syntax soup just to support variadic templates. Do you think it's
> worth the additional code complexity to do so, as opposed to a builder pattern? E.g.:

Variadic templates are  pretty standard these days, and this PR has all of the ugly syntax I expect we will need.  That `ApplyModifiers` is the end of it.

> A great example of something with a slew of unique preconditions/filters is an object rewrite
> request:
> https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite

Note that `GetBucketMetadataRequest` accepts only a subject of the well-known parameters. the code will not compile if you give it a parameter it does not know about.  **Event though** the `WellKnownParameters` class already has parameters that `GetBucketMetadataRequest` does not accept or need.  I guess what I am saying is that adding something like "Objects: rewrite" will increase the size of `WellKnownParameters` but the rest of the code, and the code that the application writes remains unchanged.

Also, this approach will allow us to pass other parameters to the request that go in headers, e.g.:

```C++
auto object = ...->CreateObject(..., UserProject("foo"), IfGenerationMatch(0),
   ContentType("image/jpeg"));
```

The application gets one function call where they can set all the optional stuff using the same syntax, and leave stuff out if they want to.

I think this is easier to *use* than a builder, if you are arguing that it is harder to write, then perhaps, but not by much, and the easy of use is an overriding goal for me.

Reply:
> the code will not compile if you give it a parameter it does not know about.
Ahhh, I see now. It forwards the parameters on through the private methods, which only accept the params that are valid for that request.  Neat.

One more variadic templates question: What about discoverability?  To be more specific... putting myself in an user's shoes, I might rely a good bit on my IDE's Intellisense-ish functionality to discover the ways in which I can modify/operate on a given type.  With the builder pattern, it's pretty straightforward -- `setUserProject(UserProject&& p)` would pop up in the dropdown list of public methods I could call on the request.  With the current approach, I'd see that the `Apply(UserProject&& p)` method is a private one, which might be a tad confusing.  With variadic templated methods like `ApplyModifiers`, would users see in that dropdown box a list of all the valid types they can pass as arguments, or do they have to dig infer from the private methods or dig through more docs to see which types from WellKnownParameters are valid?   If the former, that works for me.  (I'd check for myself, but I'm away from my workstation... on which I still need to configure CLion anyway.)

Reply:
+1 for making the individual `Apply` methods public.  For a significant portion of requests, (I suspect) there will only be one precondition added anyway, and calling `Apply(ExactType et)` directly at least saves a couple method calls.

If we're going to make them public, I'd suggest `renaming` the single `Apply` methods to `ApplyModifier` -- it's then more obvious that `ApplyModifiers` is basically shorthand for several `ApplyModifier` calls.

To that effect, the docs for `ApplyModifiers` could be simplified, saying something along the lines of "this is shorthand for `request.ApplyModifier(m1); request.ApplyModifier(m2); ...`

Reply:
> If we're going to make them public, I'd suggest renaming the single Apply methods to
> ApplyModifier -- it's then more obvious that ApplyModifiers is basically shorthand for several
> ApplyModifier calls.

Done.

> To that effect, the docs for ApplyModifiers could be simplified, saying something along the lines
> of "this is shorthand for request.ApplyModifier(m1); request.ApplyModifier(m2); ...

Done.